File: backend/app/routers/ai.py
"""``/api/ai/*`` — the proxy in front of the private ai-agent (master §6.6, §8).

Three rules hold for every endpoint in this module.

**Nothing here writes to the database (decision D-AI1).** Every response is a
*draft*. The user reviews it and the frontend then calls the ordinary
``POST /api/todos`` / ``POST /api/todos/{id}/subtasks`` / ``PATCH /api/todos/{id}``.
An LLM is not an unreviewed write path, and the rule also makes every endpoint
here idempotent and trivially testable. ``tests/test_ai_api.py`` asserts the
database is byte-identical after each call.

**Prompt context is assembled here, never in ai-agent.** ai-agent has no
database and no idea who is calling; the backend is the only component allowed
to read a user's todos, and it only ever reads *that* user's rows — the same
owner-scoped repositories every other route uses.

**Every ai-agent failure becomes one of four documented answers**: 503
``ai_disabled``, 503 ``ai_unavailable``, 504 ``ai_timeout``, 429
``rate_limited``. The user never sees an ai-agent error body, a prompt, or a
model name they did not ask for.
"""

import logging
from datetime import date, datetime, timezone
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends

from app.ai_client import AiAgentClient
from app.config import Settings
from app.db.models import Priority, Todo, User
from app.deps import (
    enforce_ai_rate_limit,
    get_ai_client,
    get_app_settings,
    get_current_user,
    get_list_repository,
    get_tag_repository,
    get_todo_repository,
)
from app.errors import AiDisabled, ListNotFound, TodoNotFound
from app.query import PRIORITY_RANK
from app.repositories.lists import SqlAlchemyListRepository
from app.repositories.protocols import TodoQuery, TodoRepository
from app.repositories.tags import SqlAlchemyTagRepository
from app.routers.utils import parse_uuid
from app.schemas.ai import (
    MAX_KNOWN_TAGS,
    MAX_SNAPSHOT_SUBTASKS,
    MAX_SUMMARY_TODOS,
    AiStatusResponse,
    DailySummaryRequest,
    DailySummaryResponse,
    EditTodoRequest,
    EditTodoResponse,
    ParseTodoRequest,
    ParseTodoResponse,
    SuggestMetadataRequest,
    SuggestMetadataResponse,
    SuggestSubtasksRequest,
    SuggestSubtasksResponse,
    TodoDraft,
)
from app.schemas.common import ErrorResponse

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/ai", tags=["ai"])

CurrentUser = Annotated[User, Depends(get_current_user)]
AppSettings = Annotated[Settings, Depends(get_app_settings)]
AiClient = Annotated[AiAgentClient, Depends(get_ai_client)]
Todos = Annotated[TodoRepository, Depends(get_todo_repository)]
Tags = Annotated[SqlAlchemyTagRepository, Depends(get_tag_repository)]
Lists = Annotated[SqlAlchemyListRepository, Depends(get_list_repository)]

AI_ERROR_RESPONSES = {
    401: {"model": ErrorResponse, "description": "Not authenticated"},
    429: {"model": ErrorResponse, "description": "Too many AI requests"},
    503: {"model": ErrorResponse, "description": "AI disabled or unavailable"},
    504: {"model": ErrorResponse, "description": "The AI service timed out"},
}
TODO_NOT_FOUND_RESPONSE = {
    404: {"model": ErrorResponse, "description": "Todo not found"}
}


def require_ai_enabled(settings: AppSettings) -> None:
    """Refuse every AI endpoint except ``/status`` when AI is switched off."""
    if not settings.AI_ENABLED:
        raise AiDisabled()


#: The dependency stack every AI POST shares, **in the order they must run**.
#:
#: Authentication first: with ``AI_ENABLED=false`` an anonymous caller used to
#: get 503 ``ai_disabled``, which answers a question it was never entitled to
#: ask — whether this deployment runs AI at all — and makes the endpoint behave
#: differently from every other authenticated route in the API. Everyone
#: unauthenticated now gets the same 401 regardless of configuration.
#:
#: Then the master switch, then the rate limit: a deployment with AI off must
#: not spend a caller's 20-per-5-minutes budget on an endpoint that was never
#: going to reach a model.
AI_GUARDS = [
    Depends(get_current_user),
    Depends(require_ai_enabled),
    Depends(enforce_ai_rate_limit),
]


async def _known_tags(tags: Tags, user_id: UUID) -> list[str]:
    """The caller's own tag vocabulary, capped at ai-agent's ``known_tags`` limit.

    Truncation is by usage rather than alphabetically: a user with 60 tags
    would otherwise silently lose everything after "m", and the tags they
    actually use are the ones worth spending prompt budget on. Sorted by name
    afterwards so the same vocabulary always produces the same prompt.
    """
    counted = await tags.list_with_counts(user_id)
    counted.sort(key=lambda row: (-row[1], row[0].name))
    return sorted(tag.name for tag, _count in counted[:MAX_KNOWN_TAGS])


async def _require_todo(todos: Todos, user_id: UUID, todo_id: str) -> Todo:
    """Resolve a body-supplied todo id within the caller's own todos.

    Unknown, malformed and *somebody else's* all collapse into the same 404
    (D1 + D-E2): an AI endpoint must not become the one place in the API where
    a caller can probe which ids exist.
    """
    parsed = parse_uuid(todo_id)
    todo = await todos.get(user_id, parsed) if parsed else None
    if todo is None:
        raise TodoNotFound()
    return todo


def _summary_sort_key(todo: Todo) -> tuple[int, date, int]:
    """Order for the summary context: due date first, then priority.

    Undated todos sort last (``1`` before the date) rather than first — a todo
    with no deadline is not more urgent than one due today, and the model reads
    the earlier entries as the more important ones.
    """
    return (
        1 if todo.due_date is None else 0,
        todo.due_date or date.max,
        PRIORITY_RANK.get(todo.priority, len(PRIORITY_RANK)),
    )


@router.get("/status", response_model=AiStatusResponse)
async def ai_status(
    _user: CurrentUser, settings: AppSettings, client: AiClient
) -> AiStatusResponse:
    """Whether the AI features can be offered. **Always 200** (master §6.6).

    This is the endpoint the UI asks before rendering anything AI-related, so
    it must answer even when every AI component is dead: ``enabled`` hides the
    section, ``available`` disables it with an explanation. An error here would
    make a broken *optional* subsystem look like a broken app.

    Authenticated like every other route, but deliberately **not** rate-limited:
    §6.6 scopes the 20-per-5-minutes budget to the POST endpoints (the ones that
    cost model time), and a 429 on the endpoint whose job is to say "AI is
    fine" would black out the UI's AI section for five minutes.

    ``reason`` (iteration 4, master §5.1) says *why* it is unavailable, as a
    closed machine enum — never prose, never a quoted upstream body. It is
    ``null`` exactly when ``available`` is true.
    """
    if not settings.AI_ENABLED:
        # Nothing is called: with AI off there is no service to ask, and the
        # answer would be the same either way.
        return AiStatusResponse(
            enabled=False, available=False, model=None, reason="disabled"
        )

    health = await client.health()
    return AiStatusResponse(
        enabled=True,
        available=health.available,
        model=health.model,
        # ``AiStatusResponse`` enforces "reason is null iff available", and this
        # endpoint may never fail — so the one combination the client half could
        # not interpret is coalesced here rather than raised at the very last
        # step of a request whose whole contract is "always 200".
        reason=health.reason or (None if health.available else "unreachable"),
    )


@router.post(
    "/parse-todo",
    response_model=ParseTodoResponse,
    dependencies=AI_GUARDS,
    responses=AI_ERROR_RESPONSES,
)
async def parse_todo(
    payload: ParseTodoRequest, user: CurrentUser, client: AiClient, tags: Tags
) -> ParseTodoResponse:
    """Turn a natural-language note into a draft todo.

    The caller sends only its text and its local date; the tag vocabulary is
    added here from the caller's own tags, so a client can neither inflate the
    prompt nor ask about tags that are not its own.
    """
    body = await client.post(
        "/ai/parse-todo",
        {
            "text": payload.text,
            "today": payload.today.isoformat(),
            "known_tags": await _known_tags(tags, user.id),
        },
    )
    # Never at info: the model's output is the user's own note, reflected.
    logger.debug("ai-agent parse-todo returned keys %s", sorted(body))
    return ParseTodoResponse(draft=TodoDraft.from_agent(body))


@router.post(
    "/suggest-subtasks",
    response_model=SuggestSubtasksResponse,
    dependencies=AI_GUARDS,
    responses={**AI_ERROR_RESPONSES, **TODO_NOT_FOUND_RESPONSE},
)
async def suggest_subtasks(
    payload: SuggestSubtasksRequest, user: CurrentUser, client: AiClient, todos: Todos
) -> SuggestSubtasksResponse:
    """Suggest the steps an existing todo breaks down into."""
    todo = await _require_todo(todos, user.id, payload.todo_id)

    body = await client.post(
        "/ai/suggest-subtasks",
        {
            "title": todo.title,
            "description": todo.description,
            "max_items": payload.max_items,
        },
    )
    return SuggestSubtasksResponse.from_agent(body, payload.max_items)


@router.post(
    "/suggest-metadata",
    response_model=SuggestMetadataResponse,
    dependencies=AI_GUARDS,
    responses={**AI_ERROR_RESPONSES, **TODO_NOT_FOUND_RESPONSE},
)
async def suggest_metadata(
    payload: SuggestMetadataRequest,
    user: CurrentUser,
    client: AiClient,
    todos: Todos,
    tags: Tags,
) -> SuggestMetadataResponse:
    """Suggest a priority and tags for an existing todo."""
    todo = await _require_todo(todos, user.id, payload.todo_id)

    body = await client.post(
        "/ai/suggest-metadata",
        {
            "title": todo.title,
            "description": todo.description,
            "known_tags": await _known_tags(tags, user.id),
            # The caller's local date when it sent one; the server's is the
            # only other option and "is this due soon" needs *a* today.
            "today": (payload.today or datetime.now(timezone.utc).date()).isoformat(),
        },
    )
    return SuggestMetadataResponse.from_agent(body)


@router.post(
    "/edit-todo",
    response_model=EditTodoResponse,
    dependencies=AI_GUARDS,
    responses={**AI_ERROR_RESPONSES, **TODO_NOT_FOUND_RESPONSE},
)
async def edit_todo(
    payload: EditTodoRequest,
    user: CurrentUser,
    client: AiClient,
    todos: Todos,
    tags: Tags,
) -> EditTodoResponse:
    """Turn a free-text instruction into a reviewable change set (§2.1).

    Like every endpoint here it **writes nothing** (D-AI1). The answer is a
    diff the user deselects parts of and confirms; the frontend then applies
    what survived through the ordinary ``PATCH /api/todos/{id}`` and subtask
    endpoints (D-IT5-1), which is also why an apply still works after the AI
    stack has died.

    What ai-agent gets is a **positional** snapshot: this todo's own fields and
    at most :data:`MAX_SNAPSHOT_SUBTASKS` subtasks as ``{title, completed}``,
    in repository order, with no identifier anywhere in it (D-IT5-3). A model
    that never sees a UUID cannot echo, mangle or invent one, and a 3B model
    copies a single digit far more reliably than 36 hex characters.

    What comes back is resolved against the real todo in
    :meth:`EditTodoResponse.from_agent` — positions become ids, no-ops are
    dropped, every value is re-clamped. That step is not bookkeeping: two
    services agreeing on a contract is not the same as this service being safe
    when the other one is wrong.
    """
    todo = await _require_todo(todos, user.id, payload.todo_id)
    if todo.parent_id is not None:
        # Decision D-IT5-6. This endpoint edits top-level todos — a snapshot of
        # a subtask has no subtasks of its own and the UI never offers the
        # action there — and D-E2 already collapses "not a valid target for
        # you" into the same 404 an unknown id gets. A distinct code here would
        # confirm that the id exists.
        raise TodoNotFound()

    # The relationship is ordered by ``(created_at, id)`` and eager-loaded by
    # the repository, so this is the same order ``TodoResponse`` shows the user
    # — which is what makes a position mean anything to them.
    ordered = list(todo.subtasks)
    snapshot = ordered[:MAX_SNAPSHOT_SUBTASKS]

    body = await client.post(
        "/ai/edit-todo",
        {
            "instruction": payload.instruction,
            "today": payload.today.isoformat(),
            "known_tags": await _known_tags(tags, user.id),
            "todo": {
                "title": todo.title,
                "description": todo.description,
                "priority": (
                    todo.priority.value
                    if isinstance(todo.priority, Priority)
                    else str(todo.priority)
                ),
                "due_date": todo.due_date.isoformat() if todo.due_date else None,
                "completed": todo.completed,
                "tags": [tag.name for tag in todo.tags],
                "subtasks": [
                    {"title": subtask.title, "completed": subtask.completed}
                    for subtask in snapshot
                ],
            },
        },
    )
    # Never at info: the reply is the user's own todo, reflected.
    logger.debug("ai-agent edit-todo returned keys %s", sorted(body))
    return EditTodoResponse.from_agent(
        body,
        todo=todo,
        subtasks=snapshot,
        truncated=len(ordered) > MAX_SNAPSHOT_SUBTASKS,
        today=payload.today,
    )


@router.post(
    "/daily-summary",
    response_model=DailySummaryResponse,
    dependencies=AI_GUARDS,
    responses={
        **AI_ERROR_RESPONSES,
        404: {"model": ErrorResponse, "description": "List not found"},
    },
)
async def daily_summary(
    payload: DailySummaryRequest,
    user: CurrentUser,
    client: AiClient,
    todos: Todos,
    lists: Lists,
) -> DailySummaryResponse:
    """A short briefing over the caller's open todos.

    Only **active top-level** todos are sent: a completed todo is not something
    to plan around, and a subtask read out of the context of its parent is
    noise. The context is capped at 50 items — enough for a real backlog, few
    enough that a 3B model writes a briefing instead of reciting a list.
    """
    scoped_list_id = None
    if payload.list_id is not None:
        # Resolved against the caller's own lists, so an unknown, malformed or
        # foreign id is one answer — 404 ``list_not_found``, the same one
        # GET /api/todos gives — rather than a summary of everything, which
        # would silently describe the wrong todos.
        parsed = parse_uuid(payload.list_id)
        owned = await lists.get(user.id, parsed) if parsed else None
        if owned is None:
            raise ListNotFound()
        scoped_list_id = owned.id

    rows, _total = await todos.list_todos(
        user.id,
        TodoQuery(
            list_id=scoped_list_id,
            status="active",
            sort="due_date",
            order="asc",
            limit=MAX_SUMMARY_TODOS,
            today=payload.today,
        ),
    )
    # The query orders by due date (undated last); priority is applied here as
    # the secondary key. Sorting ≤ 50 loaded rows in Python is cheaper than a
    # second ORDER BY term that every other caller of list_todos would pay for.
    rows.sort(key=_summary_sort_key)

    list_names = {row.id: row.name for row in await lists.list_lists(user.id)}
    context = [
        {
            "title": todo.title,
            "priority": (
                todo.priority.value
                if isinstance(todo.priority, Priority)
                else str(todo.priority)
            ),
            "due_date": todo.due_date.isoformat() if todo.due_date else None,
            "completed": todo.completed,
            "list_name": list_names.get(todo.list_id),
        }
        for todo in rows
    ]

    generated_at = datetime.now(timezone.utc)
    if not context:
        # Nothing to summarise: no model call, no waiting, no invented prose
        # about an empty list. The frontend shows its own empty copy off
        # ``todo_count == 0`` (slice-4 F5), so the contract still holds.
        return DailySummaryResponse(
            summary="", todo_count=0, generated_at=generated_at
        )

    body = await client.post(
        "/ai/daily-summary",
        {"today": payload.today.isoformat(), "todos": context},
    )
    summary = body.get("summary")
    return DailySummaryResponse(
        summary=summary if isinstance(summary, str) else "",
        todo_count=len(context),
        generated_at=generated_at,
    )

File: frontend/src/api/ai.ts
import type { Priority } from './types';
import { request } from './client';
import { ApiError } from './errors';

/**
 * The AI endpoints (master §6.6). Every one of them returns a *draft*: the AI
 * never writes to the database (decision D-AI1), so applying a suggestion goes
 * through the ordinary todo endpoints once the user has confirmed it.
 */

/**
 * The outermost rung of the timeout ladder (master §8.3): ai-agent waits 45 s
 * on Ollama, the backend 50 s on ai-agent, and the browser gives up last at
 * 60 s — so a slow model surfaces as the backend's 504 `ai_timeout` rather
 * than as our own abort, and this only fires when the backend itself hangs.
 */
export const AI_TIMEOUT_MS = 60_000;

export const AI_UNAVAILABLE_BANNER =
  'AI features are unavailable right now. Everything else works as usual.';
export const AI_AUTH_FAILED_BANNER =
  "AI is unavailable: the AI service rejected this server's credentials. Everything else works as usual.";
export const AI_MODEL_UNAVAILABLE_BANNER =
  'The AI model is still starting or has not been downloaded yet. Everything else works as usual.';
export const AI_UNAVAILABLE_MESSAGE = 'AI is unavailable right now. Please try again later.';
export const AI_RATE_LIMIT_MESSAGE = 'Too many AI requests. Please wait a moment.';
export const AI_TIMEOUT_MESSAGE = 'That took too long. Please try again.';
export const AI_SLOW_HINT = 'This can take up to a minute on a slow machine.';

/**
 * The instruction cap of decision D-IT5-9. The backend truncates at the same
 * number rather than rejecting, so this is a courtesy to the user (the textarea
 * stops accepting characters) and never the only guard.
 */
export const MAX_INSTRUCTION_LENGTH = 500;

/** Raised when our own 60 s `AbortController` fires — not a server error. */
export class AiTimeoutError extends Error {
  constructor() {
    super('The AI request took too long.');
    this.name = 'AiTimeoutError';
  }
}

/**
 * Raised when an AI edit was applied *in part* before a write failed
 * (decision D-IT5-7). Applying a change set is a sequence of ordinary calls and
 * therefore not atomic, so the panel has to be able to say how far it got —
 * and must not offer a retry, because re-running an `add` would duplicate a
 * subtask.
 */
export class PartialApplyError extends Error {
  applied: number;
  total: number;

  constructor(applied: number, total: number) {
    super(`Applied ${applied} of ${total} changes.`);
    this.name = 'PartialApplyError';
    this.applied = applied;
    this.total = total;
  }
}

/**
 * Why the AI stack is unavailable (master §5.1, iteration-4 delta). It is a
 * closed machine enum: the server promises `reason === null` exactly when
 * `available` is true, and an unrecognised value must read as `null` so the
 * server can add one later without breaking this client.
 */
export type AiUnavailableReason =
  | 'disabled'
  | 'unreachable'
  | 'auth_failed'
  | 'model_unavailable';

const AI_UNAVAILABLE_REASONS: readonly AiUnavailableReason[] = [
  'disabled',
  'unreachable',
  'auth_failed',
  'model_unavailable',
];

export interface AiStatus {
  enabled: boolean;
  available: boolean;
  model: string | null;
  /** `null` when the AI is up, when the field is missing, or when unknown. */
  reason: AiUnavailableReason | null;
}

/**
 * The banner shown next to an inert AI control. It names no host, no service
 * internals and no configuration value — only what the user can conclude from
 * it: the rest of the app is unaffected.
 *
 * `disabled` never reaches this function: `enabled === false` hides the whole
 * AI section, so there is no control left to explain.
 */
export function aiUnavailableBanner(reason: AiUnavailableReason | null | undefined): string {
  if (reason === 'auth_failed') {
    return AI_AUTH_FAILED_BANNER;
  }
  if (reason === 'model_unavailable') {
    return AI_MODEL_UNAVAILABLE_BANNER;
  }
  return AI_UNAVAILABLE_BANNER;
}

export interface AiSubtaskDraft {
  title: string;
}

export interface AiTodoDraft {
  title: string;
  description: string | null;
  priority: Priority;
  due_date: string | null;
  tags: string[];
  subtasks: AiSubtaskDraft[];
}

export interface AiMetadataSuggestion {
  priority: Priority;
  tags: string[];
}

export interface AiDailySummary {
  summary: string;
  todo_count: number;
  generated_at: string;
}

/**
 * One resolved field change of `POST /api/ai/edit-todo` (iteration-5 spec
 * §2.1). The wire key really is `from` — a Python keyword the backend aliases
 * on purpose, pinned there by a serialization test and here by a fixture.
 */
export interface AiFieldChange<T> {
  from: T;
  to: T;
}

/**
 * `to` is the full list to PATCH; `added`/`removed` are what the panel renders,
 * one checkbox per name.
 */
export interface AiTagsChange {
  from: string[];
  to: string[];
  added: string[];
  removed: string[];
}

export const AI_EDIT_ACTIONS = ['add', 'remove', 'rename', 'complete', 'reopen'] as const;

export type AiEditAction = (typeof AI_EDIT_ACTIONS)[number];

/**
 * One subtask operation. `add` carries a title and no id; `rename` carries
 * both; the other three carry an id only. `from_title` is the *server's* copy
 * of the targeted subtask's title, so the panel can label an operation without
 * trusting its own possibly-stale row.
 */
export interface AiEditOp {
  action: AiEditAction;
  id: string | null;
  title: string | null;
  from_title: string | null;
}

export interface AiEditChangeSet {
  title: AiFieldChange<string> | null;
  description: AiFieldChange<string | null> | null;
  priority: AiFieldChange<Priority> | null;
  due_date: AiFieldChange<string | null> | null;
  completed: AiFieldChange<boolean> | null;
  tags: AiTagsChange | null;
  /** Already in the order the client must apply them (§2.1). */
  subtasks: AiEditOp[];
}

export interface AiEditContext {
  /** True when the todo had more than 20 subtasks and the AI saw only the first 20. */
  subtasks_truncated: boolean;
}

export interface AiEditResult {
  todo_id: string;
  /** Derived from the parsed change set — see `editTodo`. */
  empty: boolean;
  context: AiEditContext;
  change_set: AiEditChangeSet;
}

/**
 * How many individually selectable changes a change set contains — i.e. how
 * many checkboxes the preview renders. One tag change counts once per name,
 * because that is how the user sees (and deselects) it.
 */
export function countChanges(changeSet: AiEditChangeSet): number {
  return (
    (changeSet.title === null ? 0 : 1) +
    (changeSet.description === null ? 0 : 1) +
    (changeSet.priority === null ? 0 : 1) +
    (changeSet.due_date === null ? 0 : 1) +
    (changeSet.completed === null ? 0 : 1) +
    (changeSet.tags === null ? 0 : changeSet.tags.added.length + changeSet.tags.removed.length) +
    changeSet.subtasks.length
  );
}

/** An empty change set — the shape a missing or unusable `change_set` reads as. */
export function emptyChangeSet(): AiEditChangeSet {
  return {
    title: null,
    description: null,
    priority: null,
    due_date: null,
    completed: null,
    tags: null,
    subtasks: [],
  };
}

/** True for the `AbortError` raised when the *caller* cancelled the request. */
export function isAbortError(error: unknown): boolean {
  return error instanceof DOMException && error.name === 'AbortError';
}

/** Maps a failed AI call onto the exact user-facing copy of slice-4 F5. */
export function aiErrorMessage(error: unknown): string {
  if (error instanceof AiTimeoutError) {
    return AI_TIMEOUT_MESSAGE;
  }
  if (error instanceof ApiError && error.status === 429) {
    return AI_RATE_LIMIT_MESSAGE;
  }
  // 503 `ai_unavailable` / `ai_disabled`, 504 `ai_timeout` and anything else
  // all read the same to the user: the app still works, the AI does not.
  return AI_UNAVAILABLE_MESSAGE;
}

/**
 * POSTs to an AI endpoint under a 60 s abort (master §8.3). The caller's
 * `signal` cancels it too, so unmounting or discarding a draft drops the
 * request instead of leaving it to land on a dead component.
 */
async function aiRequest<T>(url: string, body: object, signal?: AbortSignal): Promise<T> {
  const controller = new AbortController();
  let timedOut = false;

  const timer = window.setTimeout(() => {
    timedOut = true;
    controller.abort();
  }, AI_TIMEOUT_MS);

  const forwardAbort = () => controller.abort();
  if (signal?.aborted) {
    controller.abort();
  } else {
    signal?.addEventListener('abort', forwardAbort);
  }

  try {
    return await request<T>(url, 'POST', body, { signal: controller.signal });
  } catch (error) {
    if (timedOut) {
      throw new AiTimeoutError();
    }
    throw error;
  } finally {
    window.clearTimeout(timer);
    signal?.removeEventListener('abort', forwardAbort);
  }
}

function asString(value: unknown): string | null {
  return typeof value === 'string' && value.trim() !== '' ? value : null;
}

function asPriority(value: unknown): Priority {
  return value === 'low' || value === 'high' ? value : 'medium';
}

/** Accepts only the §5.1 literals; everything else — including a missing
 *  field from an older backend — reads as "no reason given". */
function asReason(value: unknown): AiUnavailableReason | null {
  return AI_UNAVAILABLE_REASONS.find((reason) => reason === value) ?? null;
}

function asTitles(value: unknown): string[] {
  if (!Array.isArray(value)) return [];
  return value
    .map((item) => asString((item as AiSubtaskDraft | null)?.title))
    .filter((title): title is string => title !== null);
}

function asTags(value: unknown): string[] {
  if (!Array.isArray(value)) return [];
  return value.filter((tag): tag is string => typeof tag === 'string' && tag !== '');
}

const ISO_DATE = /^\d{4}-\d{2}-\d{2}$/;

function asRecord(value: unknown): Record<string, unknown> | null {
  return value !== null && typeof value === 'object' && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : null;
}

/** A `{from, to}` entry, or `null` for "this field does not change" (§2.1). */
function asEntry(value: unknown): { from: unknown; to: unknown } | null {
  const record = asRecord(value);
  return record === null ? null : { from: record.from, to: record.to };
}

/** Only the three literals; unlike `asPriority` an unknown value is no change. */
function asExactPriority(value: unknown): Priority | null {
  return value === 'low' || value === 'medium' || value === 'high' ? value : null;
}

function asIsoDate(value: unknown): string | null {
  return typeof value === 'string' && ISO_DATE.test(value) ? value : null;
}

function asTitleChange(value: unknown): AiFieldChange<string> | null {
  const entry = asEntry(value);
  const to = asString(entry?.to);
  // `title` can never be cleared (§2.1), so a missing `to` is not a change.
  return entry === null || to === null
    ? null
    : { from: typeof entry.from === 'string' ? entry.from : '', to };
}

/** `description`: `to: null` means *clear it*, which is a change in itself. */
function asTextChange(value: unknown): AiFieldChange<string | null> | null {
  const entry = asEntry(value);
  if (entry === null) {
    return null;
  }
  const from = asString(entry.from);
  const to = asString(entry.to);
  // Clearing something that is already empty is not a change worth a checkbox.
  return from === null && to === null ? null : { from, to };
}

function asPriorityChange(value: unknown): AiFieldChange<Priority> | null {
  const entry = asEntry(value);
  const to = asExactPriority(entry?.to);
  return entry === null || to === null
    ? null
    : { from: asExactPriority(entry.from) ?? 'medium', to };
}

function asDueDateChange(value: unknown): AiFieldChange<string | null> | null {
  const entry = asEntry(value);
  if (entry === null) {
    return null;
  }
  const from = asIsoDate(entry.from);
  const to = asIsoDate(entry.to);
  return from === null && to === null ? null : { from, to };
}

function asCompletedChange(value: unknown): AiFieldChange<boolean> | null {
  const entry = asEntry(value);
  return entry === null || typeof entry.to !== 'boolean'
    ? null
    : { from: entry.from === true, to: entry.to };
}

function asTagsChange(value: unknown): AiTagsChange | null {
  const record = asRecord(value);
  if (record === null) {
    return null;
  }
  const added = asTags(record.added);
  const removed = asTags(record.removed);
  // `to` is what we would PATCH; without a real list a "tag change" would
  // silently clear every tag the todo has, so drop it instead.
  if (!Array.isArray(record.to) || (added.length === 0 && removed.length === 0)) {
    return null;
  }
  return { from: asTags(record.from), to: asTags(record.to), added, removed };
}

/**
 * Subtask operations, dropping anything this client could not act on: an
 * unknown `action`, a non-`add` operation without an id, and an `add` or
 * `rename` without a usable title.
 */
function asEditOps(value: unknown): AiEditOp[] {
  if (!Array.isArray(value)) {
    return [];
  }
  const ops: AiEditOp[] = [];
  for (const item of value) {
    const record = asRecord(item);
    if (record === null) {
      continue;
    }
    const action = AI_EDIT_ACTIONS.find((known) => known === record.action);
    if (action === undefined) {
      continue;
    }
    const title = asString(record.title);
    const fromTitle = asString(record.from_title);
    if (action === 'add') {
      if (title !== null) {
        ops.push({ action, id: null, title, from_title: null });
      }
      continue;
    }
    const id = asString(record.id);
    if (id === null || (action === 'rename' && title === null)) {
      continue;
    }
    ops.push({ action, id, title: action === 'rename' ? title : null, from_title: fromTitle });
  }
  return ops;
}

/**
 * `GET /api/ai/status` — the one AI endpoint that never fails. Anything other
 * than a well-formed "enabled" answer means: hide the AI section entirely.
 */
export async function getAiStatus(): Promise<AiStatus> {
  const raw = await request<Partial<AiStatus>>('/api/ai/status');
  const available = raw?.available === true;
  return {
    enabled: raw?.enabled === true,
    available,
    model: asString(raw?.model),
    // The contract guarantees `reason === null` when `available`; enforcing it
    // here means the UI can never pair a working AI with a failure banner.
    reason: available ? null : asReason(raw?.reason),
  };
}

/** `POST /api/ai/parse-todo` — a natural-language note becomes a draft todo. */
export async function parseTodo(
  text: string,
  today: string,
  signal?: AbortSignal,
): Promise<AiTodoDraft> {
  const response = await aiRequest<{ draft?: Partial<AiTodoDraft> }>(
    '/api/ai/parse-todo',
    { text, today },
    signal,
  );
  const draft = response?.draft ?? {};
  return {
    title: asString(draft.title) ?? '',
    description: asString(draft.description),
    priority: asPriority(draft.priority),
    due_date: asString(draft.due_date),
    tags: asTags(draft.tags),
    subtasks: asTitles(draft.subtasks).map((title) => ({ title })),
  };
}

/** `POST /api/ai/suggest-subtasks` — returns the suggested titles only. */
export async function suggestSubtasks(
  todoId: string,
  maxItems = 5,
  signal?: AbortSignal,
): Promise<string[]> {
  const response = await aiRequest<{ subtasks?: unknown }>(
    '/api/ai/suggest-subtasks',
    { todo_id: todoId, max_items: maxItems },
    signal,
  );
  return asTitles(response?.subtasks);
}

/** `POST /api/ai/suggest-metadata` — a priority and tags for an existing todo. */
export async function suggestMetadata(
  todoId: string,
  signal?: AbortSignal,
): Promise<AiMetadataSuggestion> {
  const response = await aiRequest<Partial<AiMetadataSuggestion>>(
    '/api/ai/suggest-metadata',
    { todo_id: todoId },
    signal,
  );
  return { priority: asPriority(response?.priority), tags: asTags(response?.tags) };
}

/**
 * `POST /api/ai/edit-todo` — a free-text instruction becomes a change set
 * (iteration-5 spec §2.1). Like every other AI endpoint it writes nothing
 * (D-AI1): the answer is a diff the user confirms, and applying it goes back
 * through the ordinary todo endpoints (D-IT5-1).
 */
export async function editTodo(
  todoId: string,
  instruction: string,
  today: string,
  signal?: AbortSignal,
): Promise<AiEditResult> {
  const response = await aiRequest<Record<string, unknown>>(
    '/api/ai/edit-todo',
    {
      todo_id: todoId,
      // The backend truncates too (D-IT5-9); doing it here as well means the
      // preview can never describe more text than we actually sent.
      instruction: instruction.slice(0, MAX_INSTRUCTION_LENGTH),
      today,
    },
    signal,
  );

  const raw = asRecord(response?.change_set) ?? {};
  const changeSet: AiEditChangeSet = {
    title: asTitleChange(raw.title),
    description: asTextChange(raw.description),
    priority: asPriorityChange(raw.priority),
    due_date: asDueDateChange(raw.due_date),
    completed: asCompletedChange(raw.completed),
    tags: asTagsChange(raw.tags),
    subtasks: asEditOps(raw.subtasks),
  };

  return {
    todo_id: asString(response?.todo_id) ?? todoId,
    // The server sends `empty` as well, and the two must agree — but what the
    // panel can actually offer is what survived parsing above, so that is what
    // decides. A server flag disagreeing with a renderable change set would
    // otherwise hide changes behind "nothing to change".
    empty: countChanges(changeSet) === 0,
    context: { subtasks_truncated: asRecord(response?.context)?.subtasks_truncated === true },
    change_set: changeSet,
  };
}

/** `POST /api/ai/daily-summary` — `listId` `null` means "across all lists". */
export async function dailySummary(
  listId: string | null,
  today: string,
  signal?: AbortSignal,
): Promise<AiDailySummary> {
  const response = await aiRequest<Partial<AiDailySummary>>(
    '/api/ai/daily-summary',
    { list_id: listId, today },
    signal,
  );
  return {
    summary: typeof response?.summary === 'string' ? response.summary : '',
    todo_count: typeof response?.todo_count === 'number' ? response.todo_count : 0,
    generated_at: asString(response?.generated_at) ?? '',
  };
}
